crates/bootstrap_srv/src/test.rs

```diff
@@ -0,0 +1,654 @@
+            Ok(res) => {
+                let res = res.into_string()?;
+                if res != "{}" {
+                    return Err(std::io::Error::other("InvalidResponse"));
+                }
+                Ok(PutInfoRes { info, agent: pk })
+            }
+            Err(ureq::Error::Status(status, res)) => {
+                let res = res.into_string()?;
+                Err(std::io::Error::other(format!("status {status}: {res}")))
+            }
+            Err(ureq::Error::Transport(err)) => Err(std::io::Error::other(err)),
+        }
+    }
+}
+
+#[test]
+fn happy_bootstrap_put_get() {
+    let s = BootstrapSrv::new(Config::testing()).unwrap();
+
+    let PutInfoRes { info, .. } = PutInfo {
+        addr: s.listen_addr(),
+        ..Default::default()
+    }
+    .call()
+    .unwrap();
+
+    let addr = format!("http://{:?}/bootstrap/{}", s.listen_addr(), S1);
+    println!("{addr}");
+    let res = ureq::get(&addr).call().unwrap().into_string().unwrap();
+    println!("{res}");
+
+    // make sure it is valid json and only contains one entry
+    let r: Vec<DecodeAgent> = serde_json::from_str(&res).unwrap();
+    assert_eq!(1, r.len());
+
+    // make sure that info byte-wise matches our put
+    assert_eq!(format!("[{info}]"), res);
+}
+
+#[test]
+fn happy_empty_server_health() {
+    let s = BootstrapSrv::new(Config::testing()).unwrap();
+    let addr = format!("http://{:?}/health", s.listen_addr());
+    let res = ureq::get(&addr).call().unwrap().into_string().unwrap();
+    assert_eq!("{}", res);
+}
+
+#[test]
+fn happy_empty_server_bootstrap_get() {
+    let s = BootstrapSrv::new(Config::testing()).unwrap();
+    let addr = format!("http://{:?}/bootstrap/{}", s.listen_addr(), S1);
+    let res = ureq::get(&addr).call().unwrap().into_string().unwrap();
+    assert_eq!("[]", res);
+}
+
+#[test]
+fn tombstone_will_not_put() {
+    let s = BootstrapSrv::new(Config::testing()).unwrap();
+
+    let _ = PutInfo {
+        addr: s.listen_addr(),
+        is_tombstone: true,
+        ..Default::default()
+    }
+    .call()
+    .unwrap();
+
+    let addr = format!("http://{:?}/bootstrap/{}", s.listen_addr(), S1);
+    let res = ureq::get(&addr).call().unwrap().into_string().unwrap();
+    assert_eq!("[]", res);
+}
+
+#[test]
+fn tombstone_old_is_ignored() {
+    let s = BootstrapSrv::new(Config::testing()).unwrap();
+
+    let _ = PutInfo {
+        addr: s.listen_addr(),
+        ..Default::default()
+    }
+    .call()
+    .unwrap();
+
+    let _ = PutInfo {
+        addr: s.listen_addr(),
+        created_at: now()
+            - std::time::Duration::from_secs(60).as_micros() as i64,
+        is_tombstone: true,
+        ..Default::default()
+    }
+    .call()
+    .unwrap();
+
+    let addr = format!("http://{:?}/bootstrap/{}", s.listen_addr(), S1);
+    let res = ureq::get(&addr).call().unwrap().into_string().unwrap();
+    let res: Vec<DecodeAgent> = serde_json::from_str(&res).unwrap();
+    assert_eq!(1, res.len());
```

Comment:
This should be 0, right? Or am I misunderstanding?

Reply:
This is a test to ensure that a PUT of an info with a created_at earlier than the one the server is tracking does not cause the tracked one to be deleted. So it should be a 1.

Reply:
ah derp